CRM/Activity/BAO/Activity.php

```diff
@@ -744,21 +743,34 @@ public static function getActivities($params) {
       'case_id' => 'case_id',
     ];
 
-    foreach ($result['values'] as $id => $activity) {
-
+    // Iterate through & do basic mappings & determine which ones we want to retrieve target count for.
+    foreach ($result as $id => $activity) {
       $activities[$id] = [];
-
-      $isBulkActivity = (!$bulkActivityTypeID || ($bulkActivityTypeID === $activity['activity_type_id']));
-      $activities[$id]['target_contact_counter'] = count($activity['target_contact_id']);
-      if ($activities[$id]['target_contact_counter']) {
-        try {
-          $activities[$id]['target_contact_name'][$activity['target_contact_id'][0]] = civicrm_api3('Contact', 'getvalue', ['id' => $activity['target_contact_id'][0], 'return' => 'sort_name']);
-        }
-        catch (CiviCRM_API3_Exception $e) {
-          // Really they should have names but a fatal here feels wrong.
-          $activities[$id]['target_contact_name'] = '';
-        }
+      $activities[$id]['activity_type_name'] = CRM_Core_PseudoConstant::getName('CRM_Activity_BAO_Activity', 'activity_type_id', $activity['activity_type_id']);
+      $activities[$id]['activity_type'] = CRM_Core_PseudoConstant::getLabel('CRM_Activity_BAO_Activity', 'activity_type_id', $activity['activity_type_id']);
+      if ($activities[$id]['activity_type_name'] === 'Bulk Email') {
+        $bulkActivities[] = $id;
+        // Get the total without permissions being passed but only display names after permissioning.
+        $activities[$id]['target_contact_count'] = civicrm_api3('ActivityContact', 'getcount', ['activity_id' => $id, 'record_type_id' => 'Activity Targets', 'check_permissions' => FALSE]);
+        $activities[$id]['recipients'] = ts('(%1 recipients)', [1 => $activities[$id]['target_contact_count']]);
+      }
+    }
+
+    // Eventually this second iteration should just handle the target contacts. It's a bit muddled at
+    // the moment as the bulk activity stuff needs unravelling & test coverage.
+    foreach ($result as $id => $activity) {
+      // Note that for the non-bulk ones we might get them in one fetch since we 'kinda know' it won't be an insane number
+      if (!isset($activities[$id]['target_contact_count']) || !empty($activities[$id]['target_contact_count'])) {
```

Comment:
I think this should be `&&` - I'm getting "Expected one ActivityContact but found 0" for contacts with activities that have no target.

Reply:
@pfigel you are right - the outer function needs the target_contact_count set even when the inner one doesn't - There is a weird (read lack of) separation of concerns here.

I've switched to getting the counts for each activity with one simple group-by query, then using a with-acl api call to get the first name for each row with one or more targets.

This could lead to a mismatch - ie they see there are '4 more contacts' but they can't 'see' who they are. I think this is the right behaviour actually